core/iwasm/aot/aot_loader.c

```diff
@@ -261,14 +293,17 @@ load_string(uint8 **p_buf, const uint8 *buf_end, AOTModule *module,
     char *str;
     uint16 str_len;
 
-    CHECK_BUF(p, p_end, 1);
-    if (*p & 0x80) {
+    p = align_ptr(p, 2);
+    if (module->const_inline_str_set != NULL &&
+        (str = bh_hash_map_find(module->const_inline_str_set, p)) != NULL) {
```

Comment:
@yamt Thanks for submitting the PR to fix the issue! It should be good, but after reviewing the source code and the AOT file format, I notice that we only repeatedly load the same string in relocation section, that means: in other sections, after the string is adjusted and returned, the loader won't load them again. In relocation section, when reading the string of section_name and symbol_name, the loader may duplicated read them, see:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/aot/aot_loader.c#L2226
and:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/aot/aot_loader.c#L2264
Here it gets the offset of the string and reads it, as the offsets may be same, the string may be loaded repeatedly.
So my understanding is that we can add the check only in `load_relocation_section`:
1. Here in `load_string`, we don't need to check whether the string is in const_inline_str_set or not, we just adjust the string and return it
2. In `load_relocation_section`, create the hash set, and before `read_string(name_addr, buf_end, group->section_name)` and `read_string(symbol_addr, buf_end, relocation->symbol_name)`, check whether `name_addr + 1`/`symbol_addr + 1` is in the hash set, if yes, directly use it, if not, load_string and add the return value in the hash set.
3. Destroy the hash set at the end of `load_relocation_section`. No need add extra hash set in aot module so as not to increase to footprint.

Reply:
while i'm not very familiar with the aot file format, (is it documented somewhere?)
i'm not aware of any other places than load_relocation_section where a string can be repeatedly loaded.
so i guess you are right.

if it's ok to rely on that level of specifics of the file format, i guess it's simpler to make validate_symbol_table "adjust" all symbols in a consistent way. that way, the extra hash set is not necessary at all. how do you think?

Reply:
Yes, that seems good, how about changing `validate_symbol_table` to `validate_and_adjust_symbol_table` or something else? And seems that we should add arguments `bool is_load_from_file_buf` to it, and inside it, handle the string according to is_load_from_file_buf: if it is true, adjust the string, else do nothing. And when read symbol string, if is_load_from_file_buf, use p+1, else use normal read_string with hash set?

Reply:
after thinking a bit more, i feel it's simpler to maintain a list of loaded symbols like https://github.com/bytecodealliance/wasm-micro-runtime/pull/1102
how do you think?

Reply:
Yes, it should be able to fix the issue, and is better and simpler.